sql/sql_insert.cc

```diff
@@ -1645,6 +1645,12 @@ int write_record(THD *thd, TABLE *table,COPY_INFO *info)
 	error= HA_ERR_FOUND_DUPP_KEY;         /* Database can't find key */
 	goto err;
       }
+	  if (table->key_info[key_nr].algorithm == HA_KEY_ALG_HASH &&
+       strcmp(table->file->table_type(), "InnoDB") == 0)
```

Comment:
db_type is legacy. Also server generally shouldn't have storage engine specific code.